model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/entities/AbstractCachedClientScope.java

```diff
@@ -0,0 +1,70 @@
+package org.keycloak.models.cache.infinispan.entities;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import org.keycloak.models.ClientScopeModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.ProtocolMapperModel;
+import org.keycloak.models.cache.infinispan.DefaultLazyLoader;
+import org.keycloak.models.cache.infinispan.LazyLoader;
+
+abstract class AbstractCachedClientScope<D extends ClientScopeModel> extends AbstractRevisioned implements InRealm {
+
+    private final LazyLoader<D, Map<String, ProtocolMapperModel>> mappersById;
+    private final LazyLoader<D, Map<String, String>> mappersByName;
+    private final LazyLoader<D, Map<String, List<String>>> mappersByType;
+
+    public AbstractCachedClientScope(Long revision, ClientScopeModel model) {
+        super(revision, model.getId());
+        mappersById = new DefaultLazyLoader<>(scope -> {
+            Map<String, ProtocolMapperModel> map = new HashMap<>();
+            scope.getProtocolMappersStream().forEach(mapper -> {
+                map.put(mapper.getId(), new ProtocolMapperModel(mapper));
+            });
+            return map;
+        }, Collections::emptyMap);
+        mappersByName = new DefaultLazyLoader<>(scope -> {
+            Map<String, String> map = new HashMap<>();
+            scope.getProtocolMappersStream().forEach(mapper -> {
+                map.put(mapper.getProtocol() + "." + mapper.getName(), mapper.getId());
+            });
+            return map;
+        }, Collections::emptyMap);
+        mappersByType = new DefaultLazyLoader<>(scope -> {
+            Map<String, List<String>> map = new HashMap<>();
+            scope.getProtocolMappersStream().forEach(mapper -> {
+                map.computeIfAbsent(mapper.getProtocolMapper(), s -> new java.util.ArrayList<>()).add(mapper.getId());
+            });
+            return map;
+        }, Collections::emptyMap);
```

Comment:
I'll let you decide if it makes things easier or harder to read. 
```suggestion
        mappersById = new DefaultLazyLoader<>(scope ->
                scope.getProtocolMappersStream()
                        .collect(Collectors.toMap(ProtocolMapperModel::getId, ProtocolMapperModel::new)), Collections::emptyMap);
        mappersByName = new DefaultLazyLoader<>(scope ->
                scope.getProtocolMappersStream()
                        .collect(Collectors.toMap(m -> m.getProtocol() + "." + m.getName(), ProtocolMapperModel::getId)), Collections::emptyMap);
        mappersByType = new DefaultLazyLoader<>(scope ->
                scope.getProtocolMappersStream()
                        .collect(Collectors.groupingBy(ProtocolMapperModel::getProtocolMapper, Collectors.mapping(ProtocolMapperModel::getId, Collectors.toList()))), Collections::emptyMap);
```